fendermint/testing/contract-test/src/lib.rs

```diff
@@ -56,3 +69,182 @@ pub async fn init_exec_state(
+                circ_supply: TokenAmount::zero(),
+                chain_id: 0,
+                power_scale: 0,
+                app_version: 0,
+            },
+        }
+    }
+
+    pub async fn init(&mut self, genesis: Genesis) -> anyhow::Result<()> {
+        let bundle_path = bundle_path();
+        let bundle = std::fs::read(&bundle_path)
+            .with_context(|| format!("failed to read bundle: {}", bundle_path.to_string_lossy()))?;
+
+        let custom_actors_bundle_path = custom_actors_bundle_path();
+        let custom_actors_bundle =
+            std::fs::read(&custom_actors_bundle_path).with_context(|| {
+                format!(
+                    "failed to read custom actors_bundle: {}",
+                    custom_actors_bundle_path.to_string_lossy()
+                )
+            })?;
+
+        let state = FvmGenesisState::new(
+            self.state_store_clone(),
+            self.multi_engine.clone(),
+            &bundle,
+            &custom_actors_bundle,
+        )
+        .await
+        .context("failed to create genesis state")?;
+
+        let (state, out) = self
+            .interpreter
+            .init(state, genesis)
+            .await
+            .context("failed to init from genesis")?;
+
+        let state_root = state.commit().context("failed to commit genesis state")?;
+
+        self.state_params = FvmStateParams {
+            state_root,
+            timestamp: out.timestamp,
+            network_version: out.network_version,
+            base_fee: out.base_fee,
+            circ_supply: out.circ_supply,
+            chain_id: out.chain_id.into(),
+            power_scale: out.power_scale,
+            app_version: 0,
+        };
+
+        Ok(())
+    }
+
+    /// Take the execution state, update it, put it back, return the output.
+    async fn modify_exec_state<T, F, R>(&self, f: F) -> anyhow::Result<T>
+    where
+        F: FnOnce(FvmExecState<MemoryBlockstore>) -> R,
+        R: Future<Output = Result<(FvmExecState<MemoryBlockstore>, T)>>,
+    {
+        let mut guard = self.exec_state.lock().await;
+        let state = guard.take().expect("exec state empty");
+
+        let (state, ret) = f(state).await?;
+
+        *guard = Some(state);
+
+        Ok(ret)
+    }
+
+    /// Put the execution state during block execution. Has to be empty.
+    async fn put_exec_state(&self, state: FvmExecState<MemoryBlockstore>) {
+        let mut guard = self.exec_state.lock().await;
+        assert!(guard.is_none(), "exec state not empty");
+        *guard = Some(state);
+    }
+
+    /// Take the execution state during block execution. Has to be non-empty.
+    async fn take_exec_state(&self) -> FvmExecState<MemoryBlockstore> {
+        let mut guard = self.exec_state.lock().await;
+        guard.take().expect("exec state empty")
+    }
+
+    pub async fn begin_block(&self, block_height: ChainEpoch) -> Result<()> {
+        // TODO: generate block hash based on input
```

Comment:
Maybe like the timestamp this could be based on the `block_height` as well, so it's not a constant, just in case something relies on its uniqueness.